updating testing-components to gjs (#2148)
ember-learn/guides-source: guides/release/testing/testing-components.md

including setting up a way to access the rendered DOM of your component later on in the test,
and cleaning up once your tests in this module are finished.

<div class="cta">
  <div class="cta-note">
    <div class="cta-note-body">
      <div class="cta-note-heading">Zoey says...</div>
      <div class="cta-note-message">
        If you generated your component using <code>ember generate component pretty-color</code> it will already have generated
        this file for you with all the boilerplate needed to get started. We are describing the steps to build to that same
        boilerplate for educational purposes. 
      </div>
    </div>
    <img src="/images/mascots/zoey.png" role="presentation" alt="Ember Mascot">
  </div>
</div>

> 

Reviewer: It might be worth noting the file extension -- I keep tripping myself up by creating files with a `.js` extension for tests and expecting `<template>` to work

});
```

Inside your `module` and after setting up the test, we can now start to create our first test case.

> Here, we can use the `QUnit.test` helper, and we can give it a descriptive name:

Reviewer: ```suggestion
Here, we can use the QUnit's `test` function, and we can give it a descriptive name:
```

    // set the outer context to red

>     this.set('colorValue', 'red');

Reviewer: Should this still be `set`?

Author: yes I tested all the examples in this whole thing... took me way longer than I had allocated to do this 🙈 🫠 

Essentially if we don't have this.set then we need to get people to initialise a class with tracked on one of the fields in every one of these tests. Maybe we might want to teach people cell (probably not) but pedagogically it's better to keep this simple and unchanged for now. We can solve this question more substantially later 👍

Reviewer: I just worry we won't revisit it since that RFC is considered part of the overall template tag RFC and we'll say that 'template tag in the guides is done'

    await render(<template> <PrettyColor @name={{this.colorValue}} /> </template>);

>     assert.equal(this.element.querySelector('div').getAttribute('style'), 'color: red', 'starts as red');

Reviewer: ```suggestion
    assert.strictEqual(this.element.querySelector('div').getAttribute('style'), 'color: red', 'starts as red');
```

    // set the outer context to red

>     this.set('colorValue', 'red');

Reviewer: Again, should these still use set?

https://rfcs.emberjs.com/id/0785-remove-set-get-in-tests

    await render(<template> <PrettyColor @name={{this.colorValue}} /> </template>);

>     assert.equal(this.element.querySelector('div').getAttribute('style'), 'color: red', 'starts as red');

Reviewer: ```suggestion
    assert.strictEqual(this.element.querySelector('div').getAttribute('style'), 'color: red', 'starts as red');
```

    this.set('colorValue', 'blue');

>     assert.equal(this.element.querySelector('div').getAttribute('style'), 'color: blue', 'updates to blue');  

Reviewer: ```suggestion
    assert.strictEqual(this.element.querySelector('div').getAttribute('style'), 'color: blue', 'updates to blue');  
```

Author: I'm trying to keep the changes as minimal as possible here 🤔 is there value in changing it to strictEqual here?

Reviewer: `assert.equal` will cause a lint error

Author: I generated a brand new Ember app using `--strict` and it doesn't have a lint error with this exact code 🤔 maybe it's something that we used to have?

Author: there is a lint error telling us not to use `@ember/runloop` 🙈 but that's a different question for a different day

Reviewer: Ah, it was, and it is only fixed in 6.8 https://github.com/ember-cli/ember-cli/pull/10707

  });
});
```

We might also test this component to ensure that the content of its template is being rendered properly:

```gjs {data-filename="tests/integration/components/pretty-color-test.gjs"}
import { module, test } from 'qunit';
import { setupRenderingTest } from 'my-app-name/tests/helpers';
import { render } from '@ember/test-helpers';
import { hbs } from 'ember-cli-htmlbars';
import PrettyColor from 'my-app-name/components/pretty-color';

module('Integration | Component | pretty-color', function(hooks) {
  setupRenderingTest(hooks);

  test('it renders', async function(assert) {
    this.set('colorValue', 'orange');

    await render(<template> <PrettyColor @name={{this.colorValue}} /> </template>);

>     assert.equal(this.element.textContent.trim(), 'Pretty Color: orange', 'text starts as orange');

Reviewer: ```suggestion
    assert.strictEqual(this.element.textContent.trim(), 'Pretty Color: orange', 'text starts as orange');
```

Reviewer: There's one below here I cannot comment suggest

  setupRenderingTest(hooks);

  test('should update title on button click', async function(assert) {
    await render(<template><MagicTitle /></template>);

>     assert.equal(this.element.querySelector('h2').textContent.trim(), 'Hello World', 'initial text is hello world');

Reviewer: ```suggestion
    assert.strictEqual(this.element.querySelector('h2').textContent.trim(), 'Hello World', 'initial text is hello world');
```

Reviewer: One below here too

>       actual = data;

Reviewer: Same question re: set

  });

  test('should change displayed location when current location changes', async function (assert) {
    await render(<template><LocationIndicator /></template>);

    assert.equal(this.element.textContent.trim(),
     'You currently are located in New York, USA', 'origin location should display');

    this.locationService = this.owner.lookup('service:location');

>     this.set('locationService.city', 'Beijing');

Reviewer: same set question -- I'm sure I'm not going to catch all of these, though
